### ISSUE TITLE
Add example of json input validation with meaningful error response

### ISSUE THREAD

**Issue by mdhtr** [mdhtr/java-webapplication#11]
Required field validation + list all missing fields is response

**Comment by mdhtr** [comment]
This issue consists of two parts, the first part is to validate the input and throw an exception with validation results, the second is to catch that exception, and return it in a good format.

For the second part, the Problem Details spec can come in handy: https://tools.ietf.org/html/rfc7807

**Comment by mdhtr** [comment]
Way A)

* no new dependencies
* custom exception class
* custom setters in the bean

1. Write a custom setter for the field you want to validate and validate the data in it before setting the field.
1. If the data is invalid, throw a custom exception that is capable of containing the validation results (eg. field name, allowed values, provided value, etc) in a way that it will be easy to extract in the exceptionmapper

**Comment by mdhtr** [comment]
Way B)

* no new dependencies
* custom exception class
* custom json deserializer

1. provide a custom deserializer to the bean
    ```java
    @JsonDeserialize(using = MyBeanDeserializer.class)
    public class MyBean {
    ...
    ```
1. implement the deserializer
    ```java
    public class MyBeanDeserializer extends JsonDeserializer {
	@Override
	public MyBean deserialize(JsonParser jp, DeserializationContext ctxt) throws IOException {
		JsonNode node = jp.getCodec().readTree(jp);
		
                ... // throw custom exception on validation error
                
		return new MyBean( ... );
	}
    ```

**Comment by mdhtr** [comment]
# Option C: JsonCreator
Tl;dr:
* custom constructor
* default Jackson exceptions

Pros:
* no new dependencies

Cons:
* very limited capacities: can only verify the presence of a field, it allows null value
---
1. Add custom constructor (example form [stackoverflow](https://stackoverflow.com/a/47807039/4456532))
```java
    @JsonCreator
    public Request(@JsonProperty(required = true) String id,
                   @JsonProperty(required = true) double code,
                   @JsonProperty(required = true) String name) {
        this.id = id;
        this.code = code;
        this.name = name;
    }
```

**Comment by mdhtr** [comment]
# Option D: custom converter + javax.validation
Tl;dr:
* custom converter
* javax.validation annotations
* javax.validation exceptions or custom exceptions

Pros:
* cleaner code

Cons:
* new dependencies
---
1. add new dependencies
    ```xml
    <!-- VALIDATION -->
        <dependency>
            <groupId>javax.validation</groupId>
            <artifactId>validation-api</artifactId>
            <version>2.0.1.Final</version>
        </dependency>
        <dependency>
            <groupId>org.hibernate.validator</groupId>
            <artifactId>hibernate-validator</artifactId>
            <version>6.0.2.Final</version>
        </dependency>
        <dependency>
            <groupId>org.hibernate.validator</groupId>
            <artifactId>hibernate-validator-annotation-processor</artifactId>
            <version>6.0.2.Final</version>
        </dependency>
        <dependency>
            <groupId>javax.el</groupId>
            <artifactId>javax.el-api</artifactId>
            <version>3.0.0</version>
        </dependency>
        <dependency>
            <groupId>org.glassfish.web</groupId>
            <artifactId>javax.el</artifactId>
            <version>2.2.6</version>
        </dependency>
    ```
1. Implement a generic validation class:
	```java
	public abstract class DataValidator<T> extends StdConverter<T, T> {
	    @Override
	    public T convert(T t) {
		ValidatorFactory validatorFactory = Validation.byDefaultProvider()
		        .configure()
		        .buildValidatorFactory();
		Validator validator = validatorFactory.getValidator();
		Set<ConstraintViolation<T>> violations = validator.validate(t);
		if (!violations.isEmpty()) {
		    throw new ConstraintViolationException(violations);
		}
		return t;
	    }
	}
	```
1. Implement type specific validation class and set it as a converter for JsonDeserialize:
	```java
	@JsonDeserialize(converter = TestValidationDto.Validator.class)
	public class TestValidationDto {
	    ...
	    static class Validator extends DataValidator<TestValidationDto> {}
	}
	```

**Comment by mdhtr** [comment]
I chose option D: custom converter + javax.validation.
The javax validation annotations make specifying validations easy, and the custom converter is highly reusable and easy to use.

**Comment by mdhtr** [comment]
A possible future improvement could be to return the validation errors is a machine parsable format, in an additional field.
